embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt

```diff
@@ -75,35 +75,50 @@
      * Loop through the payloads ready to be sent by priority and queue for delivery
      */
     private fun deliveryLoop() {
+        val failedPayloads = mutableSetOf<StoredTelemetryMetadata>()
         try {
             var deliveryQueue = createPayloadQueue()
             while (deliveryQueue.isNotEmpty() && readyToSend()) {
+                val payload = deliveryQueue.poll()
                 runCatching {
-                    deliveryQueue.poll()?.let { payload ->
-                        if (payload.shouldSendPayload() && readyToSend()) {
-                            payload.envelopeType.endpoint.updateBlockedEndpoint()
-                            queueDelivery(payload)
+                    payload?.run {
+                        if (shouldSendPayload() && readyToSend()) {
+                            envelopeType.endpoint.updateBlockedEndpoint()
+                            queueDelivery(this)
                         }
                     }
-
-                    if (queryForPayloads.compareAndSet(true, false) || deliveryQueue.isEmpty()) {
-                        deliveryQueue = createPayloadQueue()
-                    }
                 }.exceptionOrNull()?.let { error ->
-                    logger.trackInternalError(InternalErrorType.UNKNOWN_DELIVERY_ERROR, error)
+                    // This block catches unhandled errors resulting a single payload failing to be queued for delivery.
```

Comment:
Are these types of errors, which would be caused by a failure to check whether the payload associated with a StoredPayloadMetadata object should be delivered, enough to make us delete? Are these transient or not? I don't really want to do bookkeeping to see if they failed multiple times - so do we delete always or not delete always? I'm defaulting to not deleting, but I honestly can be convinced otherwise easily.